src/v3/src/components/Widget/index.tsx

```diff
@@ -327,12 +329,33 @@ export const Widget: FunctionComponent<WidgetProps> = (widgetProps) => {
 
   // bootstrap / resume the widget
   useEffect(() => {
+    /**
+     * Only for HMR at development:
+     *  If some file change triggers hot update of current file, all `useEffect` hooks will be triggered.
+     *  `useCallback` hooks will be re-executed, so values of `bootstrap` and `resume` will be updated.
+     *  But values of `useState` hooks will not update, so eg. `setIdxTransaction` value will remain.
+     *  To avoid unnecessary `bootstrap` call, compare old and new values in reduced dependencies list.
+     */
+    if (typeof __PREFRESH__ !== 'undefined') {
+      const deps: Record<string, unknown> = {
+        authClient, stateHandle, setIdxTransaction, setResponseError,
+      };
+      const prevDeps = __PREFRESH__.widgetBootstrapDeps as typeof deps;
+      const depsChanged = !prevDeps || Object.keys(prevDeps).some((k) => deps[k] !== prevDeps[k]);
+      __PREFRESH__.widgetBootstrapDeps = deps;
+      if (!depsChanged) {
+        // eslint-disable-next-line no-console
+        console.info('[HMR] Skip Widget bootstrap');
+        return;
+      }
+    }
```

Comment:
Can you extract this out into a separate file? I'd like to avoid more code in the Widget index file, especially code that is working around a development-time issue.

May even possibly even guard it with a compile-time env variable like NODE_ENV so webpack might totally strip it out in the production bundle.